File: go/300.500/480.SlidingWindowMedian.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

type hp struct {
	sort.IntSlice
	size int
}

func (h *hp) Push(v interface{}) {
	h.IntSlice = append(h.IntSlice, v.(int))
}
func (h *hp) Pop() interface{} {
	a := h.IntSlice
	v := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return v
}
func (h *hp) push(v int) {
	h.size++
	heap.Push(h, v)
}
func (h *hp) pop() int {
	h.size--
	return heap.Pop(h).(int)
}
func (h *hp) prune(d *DualHeap) {
	for h.Len() > 0 {
		num := h.IntSlice[0]
		if h == d.small {
			num = -num
		}
		if x, has := d.delayed[num]; has {
			if x > 1 {
				d.delayed[num]--
			} else {
				delete(d.delayed, num)
			}
			heap.Pop(h)
		} else {
			break
		}
	}
}

type DualHeap struct {
	delayed map[int]int
	small   *hp
	large   *hp
	k       int
}

func NewDualHeap(k int) *DualHeap {
	return &DualHeap{
		delayed: make(map[int]int),
		small:   &hp{},
		large:   &hp{},
		k:       k,
	}
}

func (d *DualHeap) makeBalance() {
	if d.small.size > d.large.size+1 {
		d.large.push(-d.small.pop())
		d.small.prune(d)
	} else if d.small.size < d.large.size {
		d.small.push(-d.large.pop())
		d.large.prune(d)
	}
}

func (d *DualHeap) insert(num int) {
	if d.small.Len() == 0 || num <= -d.small.IntSlice[0] {
		d.small.push(-num)
	} else {
		d.large.push(num)
	}
	d.makeBalance()
}

func (d *DualHeap) erase(num int) {
	d.delayed[num]++
	if num <= -d.small.IntSlice[0] {
		d.small.size--
		if num == -d.small.IntSlice[0] {
			d.small.prune(d)
		}
	} else {
		d.large.size--
		if num == d.large.IntSlice[0] {
			d.large.prune(d)
		}
	}
	d.makeBalance()
}

func (d *DualHeap) getMedian() float64 {
	if d.k&1 > 0 {
		return float64(-d.small.IntSlice[0])
	}
	return float64(-d.small.IntSlice[0]+d.large.IntSlice[0]) / 2
}

func medianSlidingWindow(nums []int, k int) []float64 {
	dp := NewDualHeap(k)
	for _, num := range nums[:k] {
		dp.insert(num)
	}
	n := len(nums)
	ans := make([]float64, 1, n-k+1)
	ans[0] = dp.getMedian()
	for i := k; i < n; i++ {
		dp.insert(nums[i])
		dp.erase(nums[i-k])
		ans = append(ans, dp.getMedian())
	}
	return ans
}

func main() {
	fmt.Println(medianSlidingWindow([]int{1, 3, -1, -3, 5, 3, 6, 7}, 3), "[1,-1,-1,3,5,6]")
	fmt.Println(medianSlidingWindow([]int{1, 2, 3, 4, 2, 3, 1, 4, 2}, 3), "[2,3,3,3,2,3,2]")
}

File: 300.500/480.SlidingWindowMedian.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
using namespace std;

/* 该对象维护两个队列
    small中所有元素小于large
    small队列元素按从小到大排列，队顶元素最大
    large队列元素按从大到小排列，队顶元素最小
    删除元素时不采用实时删除，而是当元素到大队顶时删除
    当元素不能立即删除时标记元素
    large和small的有效长度，要么相等，要么smallSize = largeSize + 1
*/
class DualHeap
{
public:
    DualHeap(int _k) : k(_k) {}

public:
    // 插入元素
    void insert(int num)
    {
        if (small.empty() || num <= small.top())
        {
            small.push(num);
            ++smallSize;
        }
        else
        {
            large.push(num);
            ++largeSize;
        }
        makeBalance();
    }
    // 删除元素
    void erase(int num)
    {
        ++delayed[num];
        if (num <= small.top())
        {
            --smallSize;
            if (num == small.top())
            {
                prune(small);
            }
        }
        else
        {
            --largeSize;
            if (num == large.top())
            {
                prune(large);
            }
        }
        makeBalance();
    }
    // 获取中位元素
    double getMedian()
    {
        if (k & 1)
        {
            // 基数长度
            return small.top();
        }
        else
        {
            // 偶数长度
            return ((double)small.top() + large.top()) / 2;
        }
    }

private:
    // 调节 small和large，使它们的长度满足需求
    void makeBalance()
    {
        if (smallSize > largeSize + 1)
        {
            large.push(small.top());
            small.pop();
            --smallSize;
            ++largeSize;
            prune(small);
        }
        else if (smallSize < largeSize)
        {
            small.push(large.top());
            large.pop();
            ++smallSize;
            --largeSize;
            prune(large);
        }
    }
    // 从优先队列移除元素，需要考虑再平衡
    template <typename T>
    void prune(T &heap)
    {
        while (!heap.empty())
        {
            int num = heap.top();
            if (delayed.count(num))
            {
                --delayed[num];
                if (!delayed[num])
                {
                    delayed.erase(num);
                }
                heap.pop();
            }
            else
            {
                break;
            }
        }
    }

private:
    // small优先队列，元素按从小到大排列，队顶元素最大
    priority_queue<int> small{};
    // large优先队列，元素按从大到小排列，队顶元素最小
    priority_queue<int, vector<int>, greater<int>> large{};
    // 记录延迟删除元素的hash表
    unordered_map<int, int> delayed{};

    // 窗口大小
    int k{};
    // small队列有效长度
    int smallSize{};
    // large队列有效长度
    int largeSize{};
};

class SlidingWindowMedian
{
public:
    vector<double> medianSlidingWindow(vector<int> nums, int k)
    {
        DualHeap dh(k);
        for (int i = 0; i < k; ++i)
        {
            dh.insert(nums[i]);
        }
        vector<double> ans(nums.size() - k + 1);
        ans[0] = dh.getMedian();
        for (int i = k; i < nums.size(); ++i)
        {
            dh.insert(nums[i]);
            dh.erase(nums[i - k]);
            ans[i - k + 1] = dh.getMedian();
        }
        return ans;
    }
};

void Print(vector<double> ret)
{
    cout << "[";
    for (auto c : ret)
    {
        cout << c << " , ";
    }
    cout << "]" << endl;
}

int main(int argc, char const *argv[])
{
    SlidingWindowMedian cls;
    // [1,-1,-1,3,5,6]
    Print(cls.medianSlidingWindow(vector<int>{1, 3, -1, -3, 5, 3, 6, 7}, 3));
    // [2,3,3,3,2,3,2]
    Print(cls.medianSlidingWindow(vector<int>{1, 2, 3, 4, 2, 3, 1, 4, 2}, 3));
    return 0;
}
